src/form/wrm.jl

```diff
@@ -76,19 +93,340 @@ function variable_voltage(pm::GenericPowerModel{T}; nw::Int=pm.cnw, cnd::Int=pm.
+        ng = length(group)
+
+        # diagonal bounds
+        for (group_idx, bus_id) in enumerate(group)
+            # group_idx indexes into group
+            # bus_id indexes into ref(pm, nw, :bus)
+            bus = ref(pm, nw, :bus, bus_id)
+
+            wr_ii = WR[group_idx, group_idx]
+
+            if bounded
+                setupperbound(wr_ii, (bus["vmax"][cnd])^2)
+                setlowerbound(wr_ii, (bus["vmin"][cnd])^2)
+            else
+                setlowerbound(wr_ii, 0)
+            end
+
+            # for non-semidefinite constraints
+            if !(bus_id in visited_buses)
+                push!(visited_buses, bus_id)
+                var(pm, nw, cnd, :w)[bus_id] = wr_ii
+            end
+        end
+
+        # off-diagonal bounds
+        offdiag_indices = [(i, j) for i in 1:ng, j in 1:ng if i != j]
+        for (i, j) in offdiag_indices
+            i_bus, j_bus = group[i], group[j]
+            if (i_bus, j_bus) in ids(pm, nw, :buspairs)
+                if bounded
+                    setupperbound(WR[i, j], wr_max[i_bus, j_bus])
+                    setlowerbound(WR[i, j], wr_min[i_bus, j_bus])
+
+                    setupperbound(WI[i, j], wi_max[i_bus, j_bus])
+                    setlowerbound(WI[i, j], wi_min[i_bus, j_bus])
+                end
+
+                # for non-semidefinite constraints
+                if !((i_bus, j_bus) in visited_buspairs)
+                    push!(visited_buspairs, (i_bus, j_bus))
+                    var(pm, nw, cnd, :wr)[(i_bus, j_bus)] = WR[i, j]
+                    var(pm, nw, cnd, :wi)[(i_bus, j_bus)] = WI[i, j]
+                end
+            end
+        end
+    end
+end
 
 ""
-function constraint_voltage(pm::GenericPowerModel{T}, nw::Int, cnd::Int) where T <: AbstractWRMForm
+function constraint_voltage(pm::GenericPowerModel{SDPWRMForm}, nw::Int, cnd::Int)
     WR = var(pm, nw, cnd)[:WR]
     WI = var(pm, nw, cnd)[:WI]
 
     @SDconstraint(pm.model, [WR WI; -WI WR] >= 0)
+end
+
+function constraint_voltage(pm::GenericPowerModel{SDPDecompForm}, nw::Int, cnd::Int)
+    pair_matrix(group) = [(i, j) for i in group, j in group]
+
+    decomp = pm.ext[:SDconstraintDecomposition]
+    groups = decomp.decomp
+    voltage_product_groups = var(pm, nw, cnd)[:voltage_product_groups]
+
+    # semidefinite constraint for each group in clique grouping
+    for (gidx, voltage_product_group) in enumerate(voltage_product_groups)
+        group = groups[gidx]
+        ng = length(group)
+        WR = voltage_product_group[:WR]
+        WI = voltage_product_group[:WI]
+
+        # Lower-dimensional SOC constraint equiv. to SDP for 2-vertex
+        # clique
+        if ng == 2
+            wr_ii = WR[1, 1]
+            wr_jj = WR[2, 2]
+            wr_ij = WR[1, 2]
+            wi_ij = WI[1, 2]
+
+            # standard SOC form (Mosek doesn't like rotated form)
+            @constraint(pm.model, (wr_ii + wr_jj)/2 >= norm([(wr_ii - wr_jj)/2; wr_ij; wi_ij]))
```

Comment:
I think this constraint is off by a factor of 2.  Double check the math.  In either case, this version should work,
```
@constraint(pm.model, (wr_ii + wr_jj) >= norm([(wr_ii - wr_jj); 2*wr_ij; 2*wi_ij]))
```

Reply:
@molzahn: We checked this against the derivation yesterday, right?